### PR TITLE
Subfolders

### LINKED ISSUE #4829: Create subfolders
As a user I want to be able to create mail folders inside other folders so that i can organize my mailbox hierarchically.

## AC
* A subfolder can be created by selecting "Add Subfolder" action next to a custom folder
* Add subfolder dialog show the parent folder as text
* Adding a new subfolder expands the parent folder.
* There is no technical limit to the depth of the nesting when creating folders
* Folder names must only be unique within the parent folder
* Newly added folders are inserted in alphabetical order
* Deleting folders with subfolders is not possible
* Folder list visualizes hierarchy and structure of folders
* Folders containing other folders can be collapsed and expanded.
* Expanded/collapsed state is restored after app restart, by default they are collapsed.
        the indentation in the folder list is limited and the tree displays as a flat list after a certain depth
* Subfolders can be (unambiguously) identified and selected from dropdown menus. (move, search, inbox rules)
        the indentation in the selection dropdown  is limited and the tree displays as a flat list after a certain depth
* it's possible to put mail in folders that contain subfolders if we display a path as text/inline, it's separated by a character/icon that's not on the keyboard
* Store new folders in a way that allows them to be retrieved in a single request

## Implementation hints
 * We should do a server-side migration which moves all the custom folders to be at the top level and not subfolders of archive (or deeper). Existing clients should be able to deal with it.
 * Do the server-side migration before releasing the client so that the parent is already set in a non-confusing way for existing folders. New client will not try load the folders recursively.
 * Currently folders have two-way references: from parent to children and from children to parent and a stored in DB in different lists, depending on the parent. This is not necessary and is inefficient
 * Store all custom folders in a single flat list together with the system folders
 * Set the parent field on the folder according to the logical structure

## Tasks
 - [x] Check that the current client will display custom folders from the top level directly
 - [ ] #4832
 - [ ] Implement new Folder Settings dialog with renaming and parent selection and use it for both "new folder" action and "edit folder" action
 - [ ] Implement new folder selection dropdown which can display the folder hierarchy
 - [ ] Change the folder list to display the hierarchy and collapsed/expanded state
 - [ ] Adapt displayed folder in search and other places

## Test notes

For migration batch job (flatten folders) please check that subfolders and inbox rules work correctly

 - [ ] Migration batch job migrates normal user mailboxes successfully
 - [ ] Migration batch job migrates shared mailboxes successfully
 - [ ] Migration batch job migrates external mailboxes (from external secure emails) successfully
 - [ ] Migration batch job migrates contact form users successfully
 - [ ] Offline db is migrated successfully (folders still load without errors)
 - [ ] Existing folders are shown as before: in a flat list in the custom folder section
 - [ ] Subfolders can be added under the custom folders
 - [ ] Unique folder name per parent folder is enforced
 - [ ] Can delete folder without mails or subfolders
 - [ ] Can't delete folder with mails
 - [ ] Can't delete folder with subfolders
 - [ ] Folders can be collapsed/expanded
 - [ ] Folders expanded state is persisted between app starts
 - [ ] Folder counters are correct
 - [ ] Folder counters of subfolders are summed up when the parent is collapsed
 - [ ] Can move emails using dropdown/shortcut
 - [ ] Can move mails using drag'n'drop
 - [ ] Folder name and path are displayed in search results
 - [ ] Can set inbox rules for subfolders
 - [ ] Inbox rules for system folders, custom folders and subfolders work

### PR DESCRIPTION
**Pull request by charlag**
close #4829

### CODE REVIEW

File: src/mail/view/MailView.ts [hunk 1]
```diff
@@ -121,36 +108,33 @@ export class MailView implements CurrentView {
...
 								? {
 									label: "newMail_action",
 									click: () => this.showNewMailDialog().catch(ofClass(PermissionError, noOp)),
 								}
 								: null,
-						content: Array.from(this.mailboxSections.entries()).map(([mailGroupId, mailboxSection]) => {
-							return m(
-								SidebarSection,
+						content: details.map((mailboxDetail) => {
+							// const mailboxSection = this.createMailboxSection(mailboxDetail)
```
Comment (ganthern) [inline src/mail/view/MailView.ts:127]: straggling comment?

File: src/mail/view/MailView.ts [hunk 2]
```diff
@@ -457,61 +432,95 @@ export class MailView implements CurrentView {
...
-		const children: Children = systemFolderButtons
-			.map(({id, button}) => {
-				const count = groupCounters[id]
-				return m(MailFolderRow, {
-					count: count,
-					button,
-					rightButton: null,
-					key: id,
-				})
+		// const [customSystems, systemSystems] = partition(mailboxDetail.folders, (f) => f.folder.folderType === MailFolderType.CUSTOM)
```
Comment (ganthern) [inline src/mail/view/MailView.ts:444]: straggling comment?

File: src/misc/DeviceConfig.ts [hunk 3]
```diff
@@ -23,7 +23,10 @@ interface ConfigObject {
 	_themeId: ThemeId
 	_language: LanguageCode | null
 	_defaultCalendarView: Record<Id, CalendarViewType | null>
+	/** map from user id to a list of calendar grouproots*/
 	_hiddenCalendars: Record<Id, Id[]>
+	/** map from user id to a list of expanded folders (elementId)*/
+	expandedMailFolders: Record<Id, Id[]>
```
Comment (ganthern) [inline src/misc/DeviceConfig.ts:29]: should this be `Record<Id, Set<Id>>`?

File: src/mail/model/FolderSystem.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import {elementIdPart, getElementId, isSameId} from "../../api/common/utils/EntityUtils.js"
+import {MailFolderType} from "../../api/common/TutanotaConstants.js"
+
+interface IndentedFolder {
+	level: number,
+	folder: MailFolder
+}
+
+/** Accessor for the folder trees. */
+export class FolderSystem {
```
Comment (ganthern) [inline src/mail/model/FolderSystem.ts:12]: I feel like this could use some tests

**Comment by ganthern** [comment]
Also needs test notes

File: src/misc/DeviceConfig.ts (hunk 3 above)
Comment (charlag) [reply src/misc/DeviceConfig.ts:29]: no, this is a plain JSON object, it can't have sets

File: src/mail/model/FolderSystem.ts (hunk 4 above)
Comment (charlag) [reply src/mail/model/FolderSystem.ts:12]: they were needed

**Pull request merged**